### DISCUSSION TITLE
How to return Table/RecordSet correctly with select+func clause

### DISCUSSION THREAD

**Discussion by zcattacz** [sqlalchemy/sqlalchemy#10242] [Usage Questions] [answered]
Take following query for example
```
rt1 = db.execute(func.pg_options_to_table(['EXPLAIN','NO_SHOW_ALL','RUNTIME_COMPUTED']))
rt2 = db.execute(text("select * from pg_options_to_table(array['EXPLAIN','NO_SHOW_ALL','RUNTIME_COMPUTED']);"))
```

In the result, `rt1` flattens the value of each row into concatenated strings.
```
2023-08-16 10:59:20,086 INFO sqlalchemy.engine.Engine SELECT pg_options_to_table(%(pg_options_to_table_2)s) AS pg_options_to_table_1
2023-08-16 10:59:20,086 INFO sqlalchemy.engine.Engine [generated in 0.00034s] {'pg_options_to_table_2': ['EXPLAIN', 'NO_SHOW_ALL', 'RUNTIME_COMPUTED']}
2023-08-16 10:59:20,105 INFO sqlalchemy.engine.Engine select * from pg_options_to_table(array['EXPLAIN','NO_SHOW_ALL','RUNTIME_COMPUTED']);
2023-08-16 10:59:20,105 INFO sqlalchemy.engine.Engine [generated in 0.00059s] {}
rt1: <sqlalchemy.engine.cursor.LegacyCursorResult object at 0x7f8719db8f70> [('(EXPLAIN,)',), ('(NO_SHOW_ALL,)',), ('(RUNTIME_COMPUTED,)',)]
rt2: <sqlalchemy.engine.cursor.LegacyCursorResult object at 0x7f8719d95ee0> [('EXPLAIN', None), ('NO_SHOW_ALL', None), ('RUNTIME_COMPUTED', None)]
```

I wonder what is the correct syntax to express `select * from func` with pattern like `select(db.func(param1, param2,...))`.

I find myself calling functions with lots of repeated typing like below, each key(ts0, ts1, etc) is repeated 3 times. Direct `db.execute(func(...))` seems definitely a no go. 
I guess there must be cleaner ways to do this, I just don't know. Can folks shed some light?

```
    stmt = text("select * from generate_report(:ts0, :ts1, :by_shift, :filter_state, :min_tsec);")\
      .bindparams(
      ts0 = para.get("ts0"), # int epoch1970 in sec
      ts1 = para.get("ts1"), # int epoch1970 in sec
      by_shift = para.get("by_shift"), #int = 1,0
      filter_state = para.get("filter_state"), #int = -1, 0, 1
      min_tsec = para.get("min_tsec")  #int
    )
```

**Comment by CaselIT** [comment]
Hi,

You are probably looking for table_valued https://docs.sqlalchemy.org/en/20/core/functions.html#sqlalchemy.sql.functions.FunctionElement.table_valued

**Comment by zcattacz** [comment]
Hi, thank you! That's a lot easier. Is it possible select the columns without specifiying them, sth like `SELECT * FROM`. I naively tried `func.NamedSqlFunc.table_valued("*")`, but the errors says it's executable.

**Reply by CaselIT** [reply, marked as answer]
You can probably use `sa.select("*").select_from(sa.func.NamedSqlFunc(...).table_valued())`

**Comment by zcattacz** [comment]
Thank you for the answer. It works just as expected. 
I looked around for `SELECT * FROM` pattern, but most of them are for ORM class based, and I didn't find a way to reflect on `func` results. 
`SELECT *` helps me focus on improving the sql function when updating the  reports (in my case, backend return list of list intead of list of dict)
